src/util/path/path.ts

```diff
@@ -297,64 +333,64 @@ export const fromArcToBeziers = (
 };
 
 /**
- * This function take a parsed SVG path and make it simpler for fabricJS logic.
- * simplification consist of: only UPPERCASE absolute commands ( relative converted to absolute )
- * S converted in C, T converted in Q, A converted in C.
- * @param {PathData} path the array of commands of a parsed svg path for `Path`
- * @return {PathData} the simplified array of commands of a parsed svg path for `Path`
+ * This function takes a parsed SVG path and makes it simpler for fabricJS logic.
+ * Simplification consist of:
+ * - All commands converted to absolute (lowercase to uppercase)
+ * - S converted to C
+ * - T converted to Q
+ * - A converted to C
+ * @param {TComplexPathData} path the array of commands of a parsed SVG path for `Path`
+ * @return {TSimplePathData} the simplified array of commands of a parsed SVG path for `Path`
+ * TODO: figure out how to remove the type assertions in a nice way
  */
-export const makePathSimpler = (path: PathData): PathData => {
-  // x and y represent the last point of the path. the previous command point.
+export const makePathSimpler = (path: TComplexPathData): TSimplePathData => {
+  // x and y represent the last point of the path, AKA the previous command point.
   // we add them to each relative command to make it an absolute comment.
   // we also swap the v V h H with L, because are easier to transform.
   let x = 0,
     y = 0;
-  const len = path.length;
   // x1 and y1 represent the last point of the subpath. the subpath is started with
   // m or M command. When a z or Z command is drawn, x and y need to be resetted to
   // the last x1 and y1.
   let x1 = 0,
     y1 = 0;
   // previous will host the letter of the previous command, to handle S and T.
   // controlX and controlY will host the previous reflected control point
-  let destinationPath: PathData = [],
-    previous,
-    controlX,
-    controlY;
-  for (let i = 0; i < len; ++i) {
-    let converted = false;
-    const current = path[i].slice(0);
+  const destinationPath: TSimplePathData = [];
+  let previous: string | undefined,
+    // placeholders
+    controlX = 0,
+    controlY = 0;
+  for (const parsedCommand of path) {
+    const current: TComplexParsedCommand = Object.assign([], parsedCommand);
+    let converted: TSimpleParsedCommand | undefined;
     switch (
       current[0] // first letter
     ) {
       case 'l': // lineto, relative
-        current[0] = 'L';
         current[1] += x;
         current[2] += y;
       // falls through
       case 'L':
         x = current[1];
         y = current[2];
+        converted = ['L', current[1], current[2]];
```

Comment:
i have read through the changes of `makePathSimpler` and i want to ask you what makes you prefer this converted array approach vs the mutation of the previous array.
Is more verbose.
current was a new array every command so is not a mutation issue. What is it?

Reply:
if we left it alone then it complains about not being able to assign 'm' to 'M' and things of that nature, which meant that either we'd need a ton of ugly casting or just plain ts-ignores. You can see the approach with types "just slapped on" more or less in the second commit.

Reply:
I don't understand this bit but casting at the top to something loose and then recasting at the bottom to something strict should do the job w/o too much ugliness

Reply:
You can always create a type util that does that for you inline I guess

Reply:
we should at least reuse the named variables where possible.
`converted = ['L', x, y]` and there are other cases like this down the file
This is added code for nothing but shouldn't be much and most of it is minifiable

Reply:
```ts
type AbsPathCommad<CmdArr extends [string,...I],I extends number[] = number[]> =
 [Uppercase<CmdArr[0]>,...I]
 
type test = AbsPathCommad<['m', 1, 3]>



function a(arg:['m',number,number]|['M',number,number]){
    switch(arg[0]){
        case 'm':
        case 'M':
        //  TS accepts this
         const upperM = (arg as AbsPathCommad<typeof arg>)[0]
            
    }
}



```



<details><summary><b>Output</b></summary>

```ts
"use strict";
function a(arg) {
    switch (arg[0]) {
        case 'm':
        case 'M':
            //  TS accepts this
            const upperM = arg[0];
    }
}

```


</details>


<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "strict": true,
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "alwaysStrict": true,
    "esModuleInterop": true,
    "declaration": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "target": "ES2017",
    "jsx": "react",
    "module": "ESNext",
    "moduleResolution": "node"
  }
}
```


</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play?#code/C4TwDgpgBAggRgZwAoENgAsDCB7AtrlAEwB5NdCYAnSqCAD2AgDtCEoBtBYSgSyYHMANADpRASQC6gsbQbNWUJgFdccCJXYSoAXkUq1GiQD4dAKA4BVMJEoBjFAgilyVDQAZjI8RPOnQkKEYuHVhEVAwcfCJidgByXFjBKABGJIBmY1Ms0wAzJSZbYB5sJigUAAoUSn4ALjiEwWVVdUb9dQkAHziAWUSmg1bmygkASgBvcygphAB3HmBbdErq9g9xyanNqHtHKHjYmo2t7YdoWN7D482AemupgBUAZTLbWwgwYDYMHgQj49sSsElNZ1N0Qst+GU2PBkGgsHgCCR-BBsDkytUjCNVj4rlcjgBfUyErJAA)

Reply:
using `as` is bad practice and should be avoided as much as possible. I really don't like this solution but I'm not set in stone, if both of you think it's better I can revert it I guess.

Reply:
Perhaps, but not using as makes us slaves to TS and its disabilities
I am for TS all along but we must draw a line when it starts driving us mad

Reply:
If it offends you using as then write a casting function (that uses as) but is written once


```ts
type AbsPathCommad<CmdArr extends [string,...I],I extends number[] = number[]> =
 [Uppercase<CmdArr[0]>,...I]
 
type test = AbsPathCommad<['m', 1, 3]>

function cast<S extends T, T = unknown>(a: T): S {
    return a as S
}


function a(arg:['m',number,number]|['M',number,number]){
    switch(arg[0]){
        case 'm':
        case 'M':
        //  TS accepts this
         const upperM = (arg as AbsPathCommad<typeof arg>)[0]
         cast<AbsPathCommad<typeof arg>>(arg)[0]
            
    }
}



```



<details><summary><b>Output</b></summary>

```ts
"use strict";
function cast(a) {
    return a;
}
function a(arg) {
    switch (arg[0]) {
        case 'm':
        case 'M':
            //  TS accepts this
            const upperM = arg[0];
            cast(arg)[0];
    }
}

```


</details>


<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "strict": true,
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "alwaysStrict": true,
    "esModuleInterop": true,
    "declaration": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "target": "ES2017",
    "jsx": "react",
    "module": "ESNext",
    "moduleResolution": "node"
  }
}
```


</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play?#code/C4TwDgpgBAggRgZwAoENgAsDCB7AtrlAEwB5NdCYAnSqCAD2AgDtCEoBtBYSgSyYHMANADpRASQC6gsbQbNWUJgFdccCJXYSoAXkUq1GiQD4dAKA4BVMJEoBjFAgilyVDQAZjI8RPOnQkKEYuHVhEVAwcfCJidgByXFjBKABGJIBmY1NTADMlJltgHmwmKHsuYgBlWUYWNgAVJLqQvIBrJmwAdyYjAAoUAC4oOoBKQaqAb3MoacoIYCVKEpQoBygK0wBfLJy8gqKlvsp+friEwWVVdXP9dQkAHziAWUSLg2vLyglhyenphA6eMBbOhDvx2B5vlNfr8ytB4rF+lDoaUHHDnojkb8APRY6Z1KooWy2CBgYBsDA8BBI5G2YrBJTWdSPEKglZseDINBYPAEEj+CDYbIrI5GYbgnyYmEOYDEDnhblRPngAVClAi3pq-hijzUyVIrZbLJAA)

Reply:
actually cast is a good util for fabric in general IMO

Reply:
> Perhaps, but not using as makes us slaves to TS and its disabilities I am for TS all along but we must draw a line when it starts driving us mad

And not damage perf for it to be happy

Reply:
Another guiding point for me I would like to point out.
The main purpose of TS is to make it easy to consume fabric.
Then to develop fabric with ease. 
So what we say is: you are aiming for ease of development in these lines but you are working harder... so the purpose is missed.
The rest of the PR you fulfill the main purpose.
TS is a tool, not a purpose on its own

Reply:
Yes in general i don't like the idea of changing code because TS doesn't like it.
TS main purpose is to help avoid errors, it helps with discovery too, but we shouldn't restrict JS at what TS understand, at least i didn't sign up for that when i decided to try a conversion.

Reply:
IDK I'm obviously biased but I still think that the current method is easier to follow. Also I'd like to point out that the refactored function is the same number of lines.

Reply:
Rejoining
I do prefer reconstructing over mutation if nothing is damaging
And here it is much rradable IMO

Reply:
So I went ahead and made it use the vars instead of `current[n]` where possible. I'm still adamant that I think this refactoring is better than doing casting, especially since it's the same number of lines. I agree that sometimes TS is overbearing but I think it's perfectly reasonable in THIS particular case

Reply:
So I think this is the last open conversation. Is this a blocker? Or are we good to go?